File: ArbolesBinariosJava/Arbol.java
public class Arbol {

    private Nodo raiz;

    public Arbol() {
        raiz = null;
    }

    public void agregar(int valor) {
        raiz = agregarRecursivo(raiz, valor);
    }

    private Nodo agregarRecursivo(Nodo actual, int numero) {
        if (actual == null) {
            return new Nodo(numero);
        }

        if (numero == actual.valor) {
            System.out.println("El número ya existe en el árbol");
            return actual;
        }

        if (numero < actual.valor) {
            actual.izquierda = agregarRecursivo(actual.izquierda, numero);
        } else {
            actual.derecha = agregarRecursivo(actual.derecha, numero);
        }

        return actual;
    }

    public void eliminar(int valor) {
        raiz = eliminarRecursivo(raiz, valor);
    }

    private Nodo eliminarRecursivo(Nodo actual, int numero) {
        if (actual == null) {
            System.out.println("El número no se encuentra en el árbol.");
            return null;
        }

        if (numero < actual.valor) {
            actual.izquierda = eliminarRecursivo(actual.izquierda, numero);
        } else if (numero > actual.valor) {
            actual.derecha = eliminarRecursivo(actual.derecha, numero);
        } else {
            if (actual.izquierda == null && actual.derecha == null) {
                return null;
            }
            if (actual.izquierda == null) {
                return actual.derecha;
            }
            if (actual.derecha == null) {
                return actual.izquierda;
            }

            Nodo sucesor = encontrarMinimo(actual.derecha);
            actual.valor = sucesor.valor;
            actual.derecha = eliminarRecursivo(actual.derecha, sucesor.valor);
        }

        return actual;
    }

    private Nodo encontrarMinimo(Nodo nodo) {
        while (nodo.izquierda != null) {
            nodo = nodo.izquierda;
        }
        return nodo;
    }

    public void mostrarPreOrden() {
        preOrden(raiz);
        System.out.println();
    }

    private void preOrden(Nodo nodo) {
        if (nodo == null) return;
        System.out.print(nodo.valor + " ");
        preOrden(nodo.izquierda);
        preOrden(nodo.derecha);
    }

    public void mostrarInOrden() {
        inOrden(raiz);
        System.out.println();
    }

    private void inOrden(Nodo nodo) {
        if (nodo == null) return;
        inOrden(nodo.izquierda);
        System.out.print(nodo.valor + " ");
        inOrden(nodo.derecha);
    }

    public void mostrarPostOrden() {
        postOrden(raiz);
        System.out.println();
    }

    private void postOrden(Nodo nodo) {
        if (nodo == null) return;
        postOrden(nodo.izquierda);
        postOrden(nodo.derecha);
        System.out.print(nodo.valor + " ");
    }
}

File: ArbolesBinariosEjem/ArbolesBinariosEjem/Arbol.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ArbolesBinariosEjem
{
    internal class Árbol
    {
        private Nodo? raiz;

        private Nodo? Raiz
        {
            get {  return raiz; }
            set { raiz = value; }
        }

        public int Numero { get; private set; }

        public Árbol()
        {
            raiz = null;
        }

        public void Agregar(int valor)
        {
            raiz = AgregarRecursivo(raiz, valor);
        }

        private Nodo AgregarRecursivo(Nodo? actual, int numero)
        {
            if (actual == null)
            {
                return new Nodo(numero);
            }
            if (numero == actual.Valor)
            {
                Console.WriteLine("El número ya existe en el árbol");
                return actual;
            }
            if (numero < actual.Valor)
            {
                actual.Izquierda = AgregarRecursivo(actual.Izquierda, numero);
            }
            else
            {
                actual.Derecha = AgregarRecursivo(actual.Derecha, numero);
            }
            return actual;
        }

        public void Eliminar(int valor) 
        { 
            raiz = EliminarRecursivo(raiz, valor); 
        }

        private Nodo? EliminarRecursivo(Nodo? actual, int? numero)
        {
            if (actual == null)
            {
                Console.WriteLine("El número no se encuentra en el árbol.");
                return null;
            }
            // Buscar el nodo
            if (numero < actual.Valor)
            {
                actual.Izquierda = EliminarRecursivo(actual.Izquierda, numero);
            }
            else if (numero > actual.Valor)
            {
                actual.Derecha = EliminarRecursivo(actual.Derecha, numero);
            }
            else
            {
                // Nodo encontrado
                // Caso 1: sin hijos
                if (actual.Izquierda == null && actual.Derecha == null)
                {
                    return null;
                }
                // Caso 2: un hijo
                if (actual.Izquierda == null)
                {
                    return actual.Derecha;
                }
                if (actual.Derecha == null)
                {
                    return actual.Izquierda;
                }
                // Caso 3: dos hijos
                // Encuentra el sucesor mínimo (menor de la derecha)
                Nodo sucesor = EncontrarMinimo(actual.Derecha);
                actual.Valor = sucesor.Valor;
                actual.Derecha = EliminarRecursivo(actual.Derecha, sucesor.Valor);
            }
            return actual;
        }

        // Método auxiliar para encontrar el nodo con el valor mínimo
        private Nodo EncontrarMinimo(Nodo nodo)
        {
            while (nodo.Izquierda != null)
            {
                nodo = nodo.Izquierda;
            }
            return nodo;
        }

        public void MostrarPreOrden()
        {
            MostrarPreOrdenRecursivo(raiz);
            Console.WriteLine();
        }

        private void MostrarPreOrdenRecursivo(Nodo? nodo)
        {
            if (nodo == null)
                return;

            // Paso 1: mostrar nodo
            Console.Write(nodo.Valor + " ");

            // Paso 2: recorrer izquierda
            MostrarPreOrdenRecursivo(nodo.Izquierda);

            // Paso 3: recorrer derecha
            MostrarPreOrdenRecursivo(nodo.Derecha);
        }

        public void MostrarInOrden()
        {
            MostrarInOrdenRecursivo(raiz);
            Console.WriteLine(); // Salto de línea al final
        }

        private void MostrarInOrdenRecursivo(Nodo? nodo)
        {
            if (nodo == null)
                return;

            // Paso 1: recorrer izquierda
            MostrarInOrdenRecursivo(nodo.Izquierda);

            // Paso 2: mostrar el valor
            Console.Write(nodo.Valor + " ");

            // Paso 3: recorrer derecha
            MostrarInOrdenRecursivo(nodo.Derecha);
        }

        public void MostrarPostOrden()
        {
            MostrarPostOrdenRecursivo(raiz);
            Console.WriteLine();
        }

        private void MostrarPostOrdenRecursivo(Nodo? nodo)
        {
            if (nodo == null)
                return;

            // Paso 1: recorrer izquierda
            MostrarPostOrdenRecursivo(nodo.Izquierda);

            // Paso 2: recorrer derecha
            MostrarPostOrdenRecursivo(nodo.Derecha);

            // Paso 3: mostrar nodo
            Console.Write(nodo.Valor + " ");
        }

    }
}
